### ISSUE TITLE
using zcat information to update numobs in mtl

### ISSUE THREAD

**Issue by forero** [desihub/desitarget#536]
The current implementation of `make_mtl` discards the zcat information when computing the `NUMOBS` that will be requested to fiberassign

https://github.com/desihub/desitarget/blob/master/py/desitarget/mtl.py#L129
```python
# ADM use passed value of NUMOBS_INIT instead of calling the memory-heavy calc_numobs.
    # ztargets['NUMOBS_MORE'] = np.maximum(0, calc_numobs(ztargets) - ztargets['NUMOBS'])
    ztargets['NUMOBS_MORE'] = np.maximum(0, targets_zmatcher['NUMOBS_INIT'] - ztargets['NUMOBS'])
```

We need to go back to using `calc_numobs` in the same way that we use `calc_priority`, i.e. to include spectroscopic information to update the MTL.

**Comment by forero** [comment]
The main case that motivates this comment is the following. Imagine starting with a target that is labeled as a `QSO` and requests `NUMOBS=4` to start with (4 is the goal for lya qsos). After the first observation we confirm that its a low redshift quasar. In this case we need to request`NUMOBS=0` and not `NUMOBS=3` as it is the current case.

**Comment by geordie666** [comment]
@forero: If you look at versions of `calc_numobs` prior to when I deprecated it, you'll see that it was never being used to set `NUMOBS=0` for LyA quasars:

https://github.com/desihub/desitarget/blob/0.31.0/py/desitarget/targets.py#L629-L630

That may have been the _intention_, but it has never been the _implementation_.

LyA quasars are set to `NUMOBS=0` based on their priorities dropping to what is appropriate for `MORE_ZGOOD`.

For example:

```
from astropy.table import Table, join
from desitarget.mtl import make_mtl
from desitarget.targetmask import desi_mask as Mx
from desitarget.targets import initial_priority_numobs
import numpy as np
# ADM set up a couple of QSO targets.
targets = Table()
types = np.array(['QSO', 'QSO'])
priorities = [Mx[t].priorities['UNOBS'] for t in types]
post_prio = [Mx[t].priorities['MORE_ZGOOD'] for t in types]
targets['DESI_TARGET'] = [Mx[t].mask for t in types]
targets['BGS_TARGET'] = np.zeros(len(types), dtype=np.int64)
targets['MWS_TARGET'] = np.zeros(len(types), dtype=np.int64)
n = len(targets)
targets['TARGETID'] = list(range(n))
pinit, ninit = initial_priority_numobs(targets)
targets["PRIORITY_INIT"] = pinit
targets["NUMOBS_INIT"] = ninit
# ADM make one of the QSOs z > 2, the other z < 2. 
zcat = Table()
zcat['TARGETID'] = targets['TARGETID']
zcat['Z'] = [2.5, 1.0]
zcat['ZWARN'] = [0, 0]
zcat['NUMOBS'] = [1, 1]
```
```
# ADM run mtl with the zcat.
mtl = make_mtl(targets, "GRAY|DARK", zcat=zcat)
```
```
Out: 
<Table length=2>
DESI_TARGET BGS_TARGET MWS_TARGET TARGETID PRIORITY_INIT NUMOBS_INIT NUMOBS_MORE PRIORITY OBSCONDITIONS
   int64      int64      int64     int64       int64        int64       int64     int64       int32    
----------- ---------- ---------- -------- ------------- ----------- ----------- -------- -------------
          4          0          0        0          3400           4           3     3500             1
          4          0          0        1          3400           4           0        2             1
```

(i.e. `NUMOBS_MORE = 3` and `NUMOBS_MORE = 0` as expected).

If you have a counter-example, please post it and I'll write a new version of `calc_numobs()`. But note that `calc_numobs()` was never, previously, changing the `NUMOBS` for quasars based on observations...in fact, that was explicitly documented in the docstring of `calc_numobs()`:

https://github.com/desihub/desitarget/blob/0.31.0/py/desitarget/targets.py#L584-L587

`calc_numobs()` was only setting the _initial_ numbers of observations for quasars, which is now done in `initial_priority_numobs()`.

**Comment by sbailey** [comment]
This appears to be a longstanding misunderstanding between `calc_numobs` explicitly saying that it isn't taking spectroscopic observations into account, and `make_mtl` thinking that it does (i.e. that `calc_numobs` is the sole decider of how many observations are needed).

Either way, @forero is right that we are currently not treating LyA vs. tracer QSOs correctly for the requested number of additional observations after the initial redshift measurement.  I will chat with @forero and @geordie666 about options and we'll report back...

**Comment by geordie666** [comment]
Just for further clarification after chatting with @forero and @sbailey: 

LyA vs. tracer quasars _were_ (I think!) being treated correctly in a broad sense. The particular cases that are failing for @forero seem to be cases where a target has two or more classes (for instance, QSOs that are also ELGs, and potentially cases that span `BRIGHT` versus `DARK|GRAY` time). @forero is going to try to workout the logic for those special cases and will push a fix.

**Comment by forero** [comment]
@geordie666 @sbailey I've just opened #537 It does solve the issues I found with DR8 data (and a made up truth table to define Lya QSOs)

**Comment by geordie666** [comment]
Addressed in #537.
